src/Parsing/Impl/Ast/FunctionScopeInfo.cs

```diff
@@ -0,0 +1,98 @@
+using Microsoft.Python.Core;
+using Microsoft.Python.Parsing.Definition;
+
+namespace Microsoft.Python.Parsing.Ast {
+    public class FunctionScopeInfo : ScopeInfo {
+        public FunctionScopeInfo(IScopeNode node) : base(node) { }
+
+        protected override bool ExposesLocalVariable => NeedsLocalsDictionary;
+
+        internal override bool TryBindOuter(IScopeNode from, string name, bool allowGlobals,
+                                            out PythonVariable variable) {
+            // Functions expose their locals to direct access
+            ContainsNestedFreeVariables = true;
+            if (TryGetVariable(name, out variable)) {
+                variable.AccessedInNestedScope = true;
+
+                if (variable.Kind == VariableKind.Local || variable.Kind == VariableKind.Parameter) {
+                    from.ScopeInfo.AddFreeVariable(variable, true);
+
+                    for (var scope = from.Parent; scope != Node; scope = scope.Parent) {
+                        scope.ScopeInfo.AddFreeVariable(variable, false);
+                    }
+
+                    AddCellVariable(variable);
+                } else if (allowGlobals) {
+                    from.ScopeInfo.AddReferencedGlobal(name);
+                }
+
+                return true;
+            }
+
+            return false;
+        }
+
+        internal override PythonVariable BindReference(PythonNameBinder binder, string name) {
+            // First try variables local to this scope
+            if (TryGetVariable(name, out var variable) && variable.Kind != VariableKind.Nonlocal) {
+                if (variable.Kind == VariableKind.Global) {
+                    AddReferencedGlobal(name);
+                }
+
+                return variable;
+            }
+
+            // Try to bind in outer scopes
+            for (var parent = Node.Parent; parent != null; parent = parent.Parent) {
+                if (parent.ScopeInfo.TryBindOuter(Node, name, true, out variable)) {
+                    return variable;
+                }
+            }
+
+            return null;
+        }
+
+
+        internal override void Bind(PythonNameBinder binder) {
+            base.Bind(binder);
+            Verify(binder);
+        }
+
+        private void Verify(PythonNameBinder binder) {
+            if (ContainsImportStar && IsClosure) {
+                binder.ReportSyntaxError(
+                                         "import * is not allowed in function '{0}' because it is a nested function"
```

Comment:
These would be good to move to Resources. Not sure why theyre hardcoded strings in the current code...

Not required, just looking.

Reply:
Ok I'll do that

Reply:
You can hold off on it for now, I may need to modify the name binder in #1586 which won't merge well.

Reply:
Though, if your PR can be done before mine, and the binding process works properly... Maybe I won't have to completely re-do scoping as well.

Reply:
I'm pretty sure the binding tests pass, I'm really just cleaning stuff up at this point unless you have any other issues.

Reply:
Actually I remember seeing some tests but I can't find them anymore, so how would I verify that the binding works?

Reply:
I have to re-do what I have anyway, so you can ignore my comments. There are some binding tests in `ParserTests.Errors`, but otherwise you'd need to search for the exact strings.

Reply:
Because comprehensions have ScopeInfo in my changes, I think the scoping errors you are trying to address can be done in name binder. I'll try to get it done in the next few days, just tell me if there's anything else besides mild styling